README.rst
apt-mirror-updater: Automated Debian/Ubuntu mirror selection
============================================================

.. image:: https://travis-ci.org/xolox/python-apt-mirror-updater.svg?branch=master
   :target: https://travis-ci.org/xolox/python-apt-mirror-updater

.. image:: https://coveralls.io/repos/xolox/python-apt-mirror-updater/badge.svg?branch=master
   :target: https://coveralls.io/r/xolox/python-apt-mirror-updater?branch=master

The `apt-mirror-updater` package automates robust apt-get_ mirror selection for
Debian_ and Ubuntu_ by enabling discovery of available mirrors, ranking of
available mirrors, automatic switching between mirrors and robust package list
updating (see features_). It's currently tested on Python 2.7, 3.5+ and PyPy
(although test coverage is still rather low, see status_).

.. contents::
   :local:

.. _features:

Features
--------

**Discovery of available mirrors**
 Debian_ and Ubuntu_ mirrors are discovered automatically by querying the
 `Debian mirror list <https://www.debian.org/mirror/list>`_ or the `Ubuntu
 mirror list <https://launchpad.net/ubuntu/+archivemirrors>`_ (the applicable
 mirror list is automatically selected based on the current platform).

**Ranking of available mirrors**
 Discovered mirrors are ranked by bandwidth (to pick the fastest mirror) and
 excluded if they're being updated (see `issues with mirror updates`_).

**Automatic switching between mirrors**
 The main mirror configured in ``/etc/apt/sources.list`` can be changed with a
 single command. The new (to be configured) mirror can be selected
 automatically or configured explicitly by the user.

**Robust package list updating**
 Several apt-get_ subcommands can fail if the current mirror is being updated
 (see `issues with mirror updates`_) and `apt-mirror-updater` tries to work
 around this by wrapping ``apt-get update`` to retry on failures and
 automatically switch to a different mirror when it looks like the current
 mirror is being updated (because I've seen such updates take more than 15
 minutes and it's not always acceptable to wait for so long, especially in
 automated solutions).

.. _status:

Status
------

On the one hand the `apt-mirror-updater` package was developed based on quite a
few years of experience in using apt-get_ on Debian_ and Ubuntu_ systems and
large scale automation of apt-get (working on 150+ remote systems). On the
other hand the Python package itself is relatively new: it was developed and
published in March 2016. As such:

.. warning:: Until `apt-mirror-updater` has been rigorously tested I consider
             it a proof of concept (beta software) so if it corrupts your
             system you can't complain that you weren't warned! I've already
             tested it on a variety of Ubuntu systems but haven't found the
             time to set up a Debian virtual machine for testing. Most of the
             logic is exactly the same though. The worst that can happen
             (assuming you trust my judgement ;-) is that
             ``/etc/apt/sources.list`` is corrupted however a backup copy is
             made before any changes are applied, so I don't see how this can
             result in irreversible corruption.

I'm working on an automated test suite but at the moment I'm still a bit fuzzy
on how to create representative tests for the error handling code paths (also,
writing a decent test suite requires a significant chunk of time :-).

Installation
------------

The `apt-mirror-updater` package is available on PyPI_ which means installation
should be as simple as:

.. code-block:: sh

   $ pip install apt-mirror-updater

There's actually a multitude of ways to install Python packages (e.g. the `per
user site-packages directory`_, `virtual environments`_ or just installing
system wide) and I have no intention of getting into that discussion here, so
if this intimidates you then read up on your options before returning to these
instructions ;-).

Usage
-----

There are two ways to use the `apt-mirror-updater` package: As the command line
program ``apt-mirror-updater`` and as a Python API. For details about the
Python API please refer to the API documentation available on `Read the Docs`_.
The command line interface is described below.

.. contents::
   :local:

.. A DRY solution to avoid duplication of the `apt-mirror-updater --help' text:
..
.. [[[cog
.. from humanfriendly.usage import inject_usage
.. inject_usage('apt_mirror_updater.cli')
.. ]]]

**Usage:** `apt-mirror-updater [OPTIONS]`

The apt-mirror-updater program automates robust apt-get mirror selection for
Debian and Ubuntu by enabling discovery of available mirrors, ranking of
available mirrors, automatic switching between mirrors and robust package list
updating.

**Supported options:**

.. csv-table::
   :header: Option, Description
   :widths: 30, 70


   "``-r``, ``--remote-host=SSH_ALIAS``","Operate on a remote system instead of the local system. The ``SSH_ALIAS``
   argument gives the SSH alias of the remote host. It is assumed that the
   remote account has root privileges or password-less sudo access."
   "``-f``, ``--find-current-mirror``","Determine the main mirror that is currently configured in
   /etc/apt/sources.list and report its URL on standard output."
   "``-b``, ``--find-best-mirror``","Discover available mirrors, rank them, select the best one and report its
   URL on standard output."
   "``-l``, ``--list-mirrors``",List available (ranked) mirrors on the terminal in a human readable format.
   "``-c``, ``--change-mirror=MIRROR_URL``",Update /etc/apt/sources.list to use the given ``MIRROR_URL``.
   "``-a``, ``--auto-change-mirror``","Discover available mirrors, rank the mirrors by connection speed and update
   status and update /etc/apt/sources.list to use the best available mirror."
   "``-u``, ``--update``, ``--update-package-lists``","Update the package lists using ""apt-get update"", retrying on failure and
   automatically switch to a different mirror when it looks like the current
   mirror is being updated."
   "``-x``, ``--exclude=PATTERN``","Add a pattern to the mirror selection blacklist. ``PATTERN`` is expected to be
   a shell pattern (containing wild cards like ""?"" and ""\*"") that is matched
   against the full URL of each mirror."
   "``-m``, ``--max=COUNT``","Don't query more than ``COUNT`` mirrors for their connection status
   (defaults to 50). If you give the number 0 no limit will be applied.
   
   Because Ubuntu mirror discovery can report more than 300 mirrors it's
   useful to limit the number of mirrors that are queried, otherwise the
   ranking of mirrors will take a long time (because over 300 connections
   need to be established)."
   "``-v``, ``--verbose``",Increase logging verbosity (can be repeated).
   "``-q``, ``--quiet``",Decrease logging verbosity (can be repeated).
   "``-h``, ``--help``",Show this message and exit.

.. [[[end]]]

.. _issues with mirror updates:

Issues with mirror updates
--------------------------

Over the past five years my team (`at work`_) and I have been managing a
cluster of 150+ Ubuntu servers, initially using manual system administration
but over time automating ``apt-get`` for a variety of use cases (provisioning,
security updates, deployments, etc.). As we increased our automation we started
running into various transient failure modes of ``apt-get``, primarily with
``apt-get update`` but incidentally also with other subcommands.

The most frequent failure that we run into is ``apt-get update`` crapping out
with 'hash sum mismatch' errors (see also `Debian bug #624122`_). When this
happens a file called ``Archive-Update-in-Progress-*`` can sometimes be found
on the index page of the mirror that is being used (see also `Debian bug
#110837`_). I've seen these situations last for more than 15 minutes.

My working theory about these 'hash sum mismatch' errors is that they are
caused by the fact that mirror updates aren't atomic, apparently causing
``apt-get update`` to download a package list whose datafiles aren't consistent
with each other. If this assumption proves to be correct (and also assuming
that different mirrors are updated at different times :-) then the command
``apt-mirror-updater --update-package-lists`` should work around this annoying
failure mode (by automatically switching to a different mirror when 'hash sum
mismatch' errors are encountered).

Publishing `apt-mirror-updater` to the world is my attempt to contribute to
this situation instead of complaining in bug trackers (see above) where no
robust and automated solution is emerging (at the time of writing). Who knows,
maybe some day these issues will be resolved by moving logic similar to what
I've implemented here into ``apt-get`` itself. Of course it would also help if
mirror updates were atomic...

Contact
-------

The latest version of `apt-mirror-updater` is available on PyPI_ and GitHub_.
The documentation is hosted on `Read the Docs`_ and includes a changelog_. For
bug reports please create an issue on GitHub_. If you have questions,
suggestions, etc. feel free to send me an e-mail at `peter@peterodding.com`_.

License
-------

This software is licensed under the `MIT license`_.

© 2020 Peter Odding.


.. External references:
.. _apt-get: https://en.wikipedia.org/wiki/Advanced_Packaging_Tool
.. _at work: http://www.paylogic.com/
.. _changelog: https://apt-mirror-updater.readthedocs.io/changelog.html
.. _Debian bug #110837: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=110837
.. _Debian bug #624122: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=624122
.. _Debian: https://en.wikipedia.org/wiki/Debian
.. _documentation: https://apt-mirror-updater.readthedocs.io
.. _GitHub: https://github.com/xolox/python-apt-mirror-updater
.. _MIT license: http://en.wikipedia.org/wiki/MIT_License
.. _per user site-packages directory: https://www.python.org/dev/peps/pep-0370/
.. _peter@peterodding.com: peter@peterodding.com
.. _PyPI: https://pypi.python.org/pypi/apt-mirror-updater
.. _Read the Docs: https://apt-mirror-updater.readthedocs.io
.. _Ubuntu: https://en.wikipedia.org/wiki/Ubuntu_(operating_system)
.. _virtual environments: http://docs.python-guide.org/en/latest/dev/virtualenvs/


CHANGELOG.rst
Changelog
=========

The purpose of this document is to list all of the notable changes to this
project. The format was inspired by `Keep a Changelog`_. This project adheres
to `semantic versioning`_.

.. contents::
   :local:

.. _Keep a Changelog: http://keepachangelog.com/
.. _semantic versioning: http://semver.org/

`Release 7.3`_ (2021-09-15)
---------------------------

**Significant changes:**

- Updated the releases bundled in the :mod:`~apt_mirror_updater.releases`
  module to include the following:

  - Ubuntu 20.10 (Groovy Gorilla)
  - Ubuntu 21.04 (Hirsute Hippo)
  - Ubuntu 21.10 (Impish Indri)

- Relaxed the :pypi:`beautifulsoup4` requirement to facilitate Python 2
  compatibility (release 4.9.3 was the last to support Python 2 whereas
  our requirements insisted on >= 4.4.1).

**Miscellaenous changes:**

- Updated the :func:`~apt_mirror_updater.releases.parse_version()` documentation
  to refer to :class:`decimal.Decimal` instead of "a floating point number" (to
  avoid any confusion).

.. _Release 7.3: https://github.com/xolox/python-apt-mirror-updater/compare/7.2...7.3

`Release 7.2`_ (2020-04-18)
---------------------------

Add support for `Elementary OS`_ (suggested in `issue #10`_).

Because Elementary OS is based on Ubuntu the new backend is nothing more than a
thin wrapper for the Ubuntu backend. Its most significant content is a data
structure with known Elementary OS releases and how they map to Ubuntu.

.. _Release 7.2: https://github.com/xolox/python-apt-mirror-updater/compare/7.1...7.2
.. _Elementary OS: https://en.wikipedia.org/wiki/Elementary_OS
.. _issue #10: https://github.com/xolox/python-apt-mirror-updater/issues/10

`Release 7.1`_ (2020-04-17)
---------------------------

After Ubuntu 14.04 went EOL it became clear that when a release goes EOL and
when it is archived (moved from the package mirrors to the old-releases
environment) can differ by a year...

This release of apt-mirror-updater acknowledges this distinction in the code
and properly handles the situation where a release has already gone EOL but has
not yet been archived.

This was also reported in `issue #9`_ by a colleague and friend of mine.

.. _Release 7.1: https://github.com/xolox/python-apt-mirror-updater/compare/7.0...7.1
.. _issue #9: https://github.com/xolox/python-apt-mirror-updater/issues/9

`Release 7.0`_ (2020-04-16)
---------------------------

This is a maintenance release that (primarily) updates Python compatibility.

**Backwards incompatible changes:**

- Python 3.7 and 3.8 are now officially supported.

- Python 2.6 and 3.4 are no longer supported.

- Added ``python_requires`` to ``setup.py`` to aid :pypi:`pip` in version
  selection given these compatibility changes.

**Significant changes:**

- Updated the releases bundled in the :mod:`apt_mirror_updater.releases` module
  to include the following:

  - Ubuntu 19.04 (Disco Dingo)
  - Ubuntu 19.10 (Eoan Ermine)
  - Ubuntu 20.04 (Focal Fossa).

- Bug fix for "accidental tuple" in :func:`apt_mirror_updater.releases.parse_csv_file()`.

**Miscellaenous changes:**

- Spent some time stabilizing the test suite on Travis CI (tests were passing
  for me locally but not on Travis CI because the mirror selection differed).
  As a result the test suite got a bit slower, but it's not too bad.

- Move caching decorator to :pypi:`humanfriendly`.

- Fixed deprecation warnings emitted by recent :pypi:`humanfriendly` releases
  and bumped requirements I authored that went through similar changes.

- Made :mod:`multiprocessing` usage compatible with coverage collection. Note
  that I don't expect this to increase coverage, I just wanted to get rid of
  the warnings 😇 (because warnings about harmless things are just as
  distracting as more pertinent warnings).

- Default to Python 3 for local development (required by :pypi:`Sphinx` among
  other things).

- Fixed existing :pypi:`Sphinx` reference warnings in the documentation and
  changed the :man:`sphinx-build` invocation to promote warnings to errors (to
  aid me in the discipline of not introducing broken references from now on).

.. _Release 7.0: https://github.com/xolox/python-apt-mirror-updater/compare/6.1...7.0

`Release 6.1`_ (2018-10-19)
---------------------------

- Bug fix for Ubuntu keyring selection that prevented
  ``ubuntu-archive-removed-keys.gpg`` from being used.
- Bug fix for :func:`~apt_mirror_updater.releases.coerce_release()`
  when given a release number.
- Moved pathnames of Debian and Ubuntu keyring files to constants.
- Added logging to enable debugging of keyring selection process.
- Added proper tests for keyring selection and release coercion.

.. _Release 6.1: https://github.com/xolox/python-apt-mirror-updater/compare/6.0...6.1

`Release 6.0`_ (2018-10-14)
---------------------------

Enable the creation of Ubuntu <= 12.04 chroots on Ubuntu >= 17.04 hosts by
working around (what I am convinced is) a bug in :man:`debootstrap` which picks
the wrong keyring when setting up chroots of old releases. For more information
refer to issue `#8`_.

I've bumped the major version number for this release because the highly
specific ``apt_mirror_updater.eol`` module changed into the much more generic
:mod:`apt_mirror_updater.releases` module. Also the ``release_label`` property was
removed.

.. _Release 6.0: https://github.com/xolox/python-apt-mirror-updater/compare/5.2...6.0
.. _#8: https://github.com/xolox/python-apt-mirror-updater/issues/8

`Release 5.2`_ (2018-10-08)
---------------------------

Use `mirrors.ubuntu.com/mirrors.txt`_ without placing our full trust in it like
older versions of :pypi:`apt-mirror-updater` did 😇.

Feedback in issue `#6`_ suggested that `mirrors.ubuntu.com/mirrors.txt`_ is
working properly (again) and should be preferred over scraping Launchpad.
However I prefer for :pypi:`apt-mirror-updater` to be a reliable "do what I
mean" program and `mirrors.ubuntu.com/mirrors.txt`_ has proven to be unreliable

apt_mirror_updater/cli.py
# Automated, robust apt-get mirror selection for Debian and Ubuntu.
#
# Author: Peter Odding <peter@peterodding.com>
# Last Change: April 15, 2020
# URL: https://apt-mirror-updater.readthedocs.io

"""
Usage: apt-mirror-updater [OPTIONS]

The apt-mirror-updater program automates robust apt-get mirror selection for
Debian and Ubuntu by enabling discovery of available mirrors, ranking of
available mirrors, automatic switching between mirrors and robust package list
updating.

Supported options:

  -r, --remote-host=SSH_ALIAS

    Operate on a remote system instead of the local system. The SSH_ALIAS
    argument gives the SSH alias of the remote host. It is assumed that the
    remote account has root privileges or password-less sudo access.

  -f, --find-current-mirror

    Determine the main mirror that is currently configured in
    /etc/apt/sources.list and report its URL on standard output.

  -b, --find-best-mirror

    Discover available mirrors, rank them, select the best one and report its
    URL on standard output.

  -l, --list-mirrors

    List available (ranked) mirrors on the terminal in a human readable format.

  -c, --change-mirror=MIRROR_URL

    Update /etc/apt/sources.list to use the given MIRROR_URL.

  -a, --auto-change-mirror

    Discover available mirrors, rank the mirrors by connection speed and update
    status and update /etc/apt/sources.list to use the best available mirror.

  -u, --update, --update-package-lists

    Update the package lists using `apt-get update', retrying on failure and
    automatically switch to a different mirror when it looks like the current
    mirror is being updated.

  -x, --exclude=PATTERN

    Add a pattern to the mirror selection blacklist. PATTERN is expected to be
    a shell pattern (containing wild cards like `?' and `*') that is matched
    against the full URL of each mirror.

  -m, --max=COUNT

    Don't query more than COUNT mirrors for their connection status
    (defaults to 50). If you give the number 0 no limit will be applied.

    Because Ubuntu mirror discovery can report more than 300 mirrors it's
    useful to limit the number of mirrors that are queried, otherwise the
    ranking of mirrors will take a long time (because over 300 connections
    need to be established).

  -v, --verbose

    Increase logging verbosity (can be repeated).

  -q, --quiet

    Decrease logging verbosity (can be repeated).

  -h, --help

    Show this message and exit.
"""

# Standard library modules.
import functools
import getopt
import logging
import sys

# External dependencies.
import coloredlogs
from executor.contexts import LocalContext, RemoteContext
from humanfriendly import format_size, format_timespan
from humanfriendly.tables import format_smart_table
from humanfriendly.terminal import connected_to_terminal, output, usage, warning

# Modules included in our package.
from apt_mirror_updater import MAX_MIRRORS, AptMirrorUpdater

# Initialize a logger for this module.
logger = logging.getLogger(__name__)


def main():
    """Command line interface for the ``apt-mirror-updater`` program."""
    # Initialize logging to the terminal and system log.
    coloredlogs.install(syslog=True)
    # Command line option defaults.
    context = LocalContext()
    updater = AptMirrorUpdater(context=context)
    limit = MAX_MIRRORS
    actions = []
    # Parse the command line arguments.
    try:
        options, arguments = getopt.getopt(sys.argv[1:], 'r:fblc:aux:m:vqh', [
            'remote-host=', 'find-current-mirror', 'find-best-mirror',
            'list-mirrors', 'change-mirror', 'auto-change-mirror', 'update',
            'update-package-lists', 'exclude=', 'max=', 'verbose', 'quiet',
            'help',
        ])
        for option, value in options:
            if option in ('-r', '--remote-host'):
                if actions:
                    msg = "The %s option should be the first option given on the command line!"
                    raise Exception(msg % option)
                context = RemoteContext(value)
                updater = AptMirrorUpdater(context=context)
            elif option in ('-f', '--find-current-mirror'):
                actions.append(functools.partial(report_current_mirror, updater))
            elif option in ('-b', '--find-best-mirror'):
                actions.append(functools.partial(report_best_mirror, updater))
            elif option in ('-l', '--list-mirrors'):
                actions.append(functools.partial(report_available_mirrors, updater))
            elif option in ('-c', '--change-mirror'):
                actions.append(functools.partial(updater.change_mirror, value))
            elif option in ('-a', '--auto-change-mirror'):
                actions.append(updater.change_mirror)
            elif option in ('-u', '--update', '--update-package-lists'):
                actions.append(updater.smart_update)
            elif option in ('-x', '--exclude'):
                actions.insert(0, functools.partial(updater.ignore_mirror, value))
            elif option in ('-m', '--max'):
                limit = int(value)
            elif option in ('-v', '--verbose'):
                coloredlogs.increase_verbosity()
            elif option in ('-q', '--quiet'):
                coloredlogs.decrease_verbosity()
            elif option in ('-h', '--help'):
                usage(__doc__)
                return
            else:
                assert False, "Unhandled option!"
        if not actions:
            usage(__doc__)
            return
        # Propagate options to the Python API.
        updater.max_mirrors = limit
    except Exception as e:
        warning("Error: Failed to parse command line arguments! (%s)" % e)
        sys.exit(1)
    # Perform the requested action(s).
    try:
        for callback in actions:
            callback()
    except Exception:
        logger.exception("Encountered unexpected exception! Aborting ..")
        sys.exit(1)


def report_current_mirror(updater):
    """Print the URL of the currently configured ``apt-get`` mirror."""
    output(updater.current_mirror)


def report_best_mirror(updater):
    """Print the URL of the "best" mirror."""
    output(updater.best_mirror)


def report_available_mirrors(updater):
    """Print the available mirrors to the terminal (in a human friendly format)."""
    if connected_to_terminal():
        have_bandwidth = any(c.bandwidth for c in updater.ranked_mirrors)
        have_last_updated = any(c.last_updated is not None for c in updater.ranked_mirrors)
        column_names = ["Rank", "Mirror URL", "Available?", "Updating?"]
        if have_last_updated:
            column_names.append("Last updated")
        if have_bandwidth:
            column_names.append("Bandwidth")
        data = []
        for i, candidate in enumerate(updater.ranked_mirrors, start=1):
            row = [i, candidate.mirror_url,
                   "Yes" if candidate.is_available else "No",
                   "Yes" if candidate.is_updating else "No"]
            if have_last_updated:
                row.append("Up to date" if candidate.last_updated == 0 else (
                    "%s behind" % format_timespan(candidate.last_updated)
                    if candidate.last_updated else "Unknown"
                ))
            if have_bandwidth:
                row.append("%s/s" % format_size(round(candidate.bandwidth, 2))
                           if candidate.bandwidth else "Unknown")
            data.append(row)
        output(format_smart_table(data, column_names=column_names))
    else:
        output(u"\n".join(
            candidate.mirror_url for candidate in updater.ranked_mirrors
            if candidate.is_available and not candidate.is_updating
        ))
